### PR TITLE
added Formatter\Presenter\Differ\ArrayEngine

### PR DESCRIPTION
**Pull request by liuggio**
I created the `ArrayEngine`, it is very similar to the `StringEngine`
then I added to the application.

I  did a lot of reverse eng. so let me know if I  miss something.
If everything is ok, I'll procede with the `ObjectEngine`

### CODE REVIEW

**Comment by everzet** [comment]
Could you add an examples (spec examples) for complex deep arrays and arrays with different depths diffing. I actually was thinking about custom recursive differ, but if PHP_Diff is capable to produce meaningful diffs for arrays - why not. We just need to check :)

**Comment by liuggio** [comment]
yes, totally right,
at first glance, seems that PHP_diff doesn't handle associative arrays, neither nested.
For the associative we could first order the array by keys then flatten,
  how to flatten: `$key . " => " . $value` ?
For the recursive I don't see any problem for now.

**Comment by everzet** [comment]
I don't think we need classical DIFF here. I would love to have output like:

```
1. item1 > subitem1 > option1 is "...", but expected to be "..."
2. item1 > subitem2 does not exists, but expected to be
3. item1 > subitem3 exists, but expected not to be
```

What do you think?

**Comment by karolsojko** [comment]
I would like the key to be included in this one as well. Many times I have runned into a situation that the keys weren't preserved by some sorting function or so. This lead to quite some time debuging that the contents of the array are identical but the keys differ. 

In general I'm :+1: for this as hell :)

**Comment by liuggio** [comment]
OK, your idea contain more info.

I'm try to understand how to play with the diff in a real example,
I lost something i thought that the diff string function handles particulars 'shouldbe' but 
 for example in my personal library,
I'm not able to raise the execution of the function the Differ:compare() 
I need some tips here.
thanks

**Comment by liuggio** [comment]
At least, I understand there's the verbose mode, that enables the comparison, hurray I can continue this task.

**Comment by everzet** [comment]
:)

**Comment by liuggio** [comment]
For sure this could be improved,
and this is not the final version,

there are some smells:
1. English text without a translation (the Symfony2 translation components is great :) )
2. Variables with very bad name (in the next commits I'll try to clean a little) and numbers without constant
3. makeResponse should be refactored (without the domain)

let me know your impressions

**Comment by liuggio** [comment]
I cleaned a little bit.

**Comment by liuggio** [comment]
squashed, for me this is finished, let me know.

**Comment by everzet** [comment]
I don't really like the syntax. It's not very intuitive :/

How about changing:

```
a is "first", but expected to be "second"
```

to:

```
$a["a"] is "first", but expected to be "second"
```

and:

```
a1->5->a3 a4b is "102", but expected to be "104"
```

to become:

```
$a["a1"][5]["a3"]["a4b"] is "102", but expected to be "104"
```

What do you think? I think it'd be much more readable and clear for newcomers.

**Comment by liuggio** [comment]
Sure is more readable, but $a doesn't have any sense to me, maybe $expected :) 
is very easy to do in 2 min I think

`$expected["0"] is "b", but expected to be "d"`
is not correct should be `$expected[0]  ...`

**Comment by liuggio** [comment]
:+1:

**Comment by everzet** [comment]
Change $expected to $actual and we're done :)

**Comment by everzet** [comment]
And one more thing - use `PHPSpec2\Formatter\Presenter\PresenterInterface::presentValue(...)` (`value_presenter` service in container) to put item values into output string. Cuz your `sprintf("%s ... %s", $expected, $actual)` will throw exception in case if `$expected` OR `$actual` are objects.

And you don't need to quote values in string if you're using `PresenterInterface` - he'll do that for you automatically (if value is a string).

File: src/PHPSpec2/Formatter/Presenter/Differ/ArrayEngine.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
+<?php
+
+namespace PHPSpec2\Formatter\Presenter\Differ;
+
+class ArrayEngine implements EngineInterface
+{
+    CONST ARRAY_VARIABLE_NAME = '$expected';
+    CONST VALUE_NOT_EXPECTED_TO_BE = 1;
+    CONST VALUE_EXPECTED_TO_BE = 2;
+    CONST VALUE_EXPECTED_TO_BE_DIFFERENT = 3;
```
Comment (everzet) [inline src/PHPSpec2/Formatter/Presenter/Differ/ArrayEngine.php:10]: `CONST` is a keyword and should be lowercased by convention :)

**Comment by liuggio** [comment]
I've got a problem with the Service Locator
adding this

``` php
class ArrayEngine implements EngineInterface
{
    private $presenter;

    public function __construct(PresenterInterface $presenter)
    {
        $this->presenter = $presenter;
    }

```

and also just injecting

``` php

         $c->extend('differ.engines',
             $c->set('differ.engines.array', $c->share(function($c) {
                return new Presenter\Differ\ArrayEngine($c('value_presenter'));
             }))
         );
```

I've got this error:

``` sh

$ bin/phpspec run -v
PHP Fatal error:  Maximum function nesting level of '100' reached, aborting! in /var/www/phpspec2/src/PHPSpec2/ServiceContainer.php on line 20
PHP Stack trace:
PHP   1. {main}() /var/www/phpspec2/bin/phpspec:0
PHP   2. Symfony\Component\Console\Application->run() /var/www/phpspec2/bin/phpspec:23
PHP   3. PHPSpec2\Console\Application->doRun() /var/www/phpspec2/vendor/symfony/console/Symfony/Component/Console/Application.php:105
PHP   4. Symfony\Component\Console\Application->doRun() /var/www/phpspec2/src/PHPSpec2/Console/Application.php:237
PHP   5. Symfony\Component\Console\Command\Command->run() /var/www/phpspec2/vendor/symfony/console/Symfony/Component/Console/Application.php:192
PHP   6. PHPSpec2\Console\Command\RunCommand->execute() /var/www/phpspec2/vendor/symfony/console/Symfony/Component/Console/Command/Command.php:238
PHP   7. PHPSpec2\ServiceContainer->__invoke() /var/www/phpspec2/src/PHPSpec2/Console/Command/RunCommand.php:48
...
...
...

```

**Comment by everzet** [comment]
Damn. Because `value_presenter` depends on `differ`, which depends on all engines one of which (yours) now depends on `value_presenter`. So it's a circular dependency:
https://github.com/phpspec/phpspec2/blob/develop/src/PHPSpec2/Console/Application.php#L80 :/

**Comment by everzet** [comment]
We probably need to move `presentException()` out of `PresenterInterface` now :)

**Comment by everzet** [comment]
@liuggio this task become more and more "fun" every minute :)

**Comment by liuggio** [comment]
:)
Ok I think is better to close this task, then create 2 issues, the first about moving `presentException()` and the other about inserting a dependency into ArrayEngine

So we can agility continue a little step at time.

I'm going to fix only little problems

What do you think?

**Comment by everzet** [comment]
:+1:

**Pull request closed without merging**